### PR TITLE
A bunch of tweaks to `ClientHelpers`' methods

### PR DESCRIPTION
**Pull request by danicheg**
This contains
* backport of https://github.com/http4s/http4s/pull/7167 to 0.23-series
* fixes https://github.com/http4s/http4s/issues/7108
* other things from https://github.com/http4s/http4s/pull/7170

### CODE REVIEW

File: ember-client/shared/src/main/scala/org/http4s/ember/client/internal/ClientHelpers.scala [hunk 1]
```diff
@@ -295,4 +292,11 @@ private[client] object ClientHelpers {
         case _ => false
       }
   }
+
+  private case class MissingHost() extends RuntimeException("Invalid hostname")
+
+  private case class MissingPort() extends RuntimeException("Invalid port")
```
Comment (armanbilge) [inline ember-client/shared/src/main/scala/org/http4s/ember/client/internal/ClientHelpers.scala:298]: Should we include the actual value, to help with debugging the error?

Comment (danicheg) [reply ember-client/shared/src/main/scala/org/http4s/ember/client/internal/ClientHelpers.scala:298]: I thought about that, but left it without actual values like Christopher has done in `H2Client`.

Comment (armanbilge) [reply ember-client/shared/src/main/scala/org/http4s/ember/client/internal/ClientHelpers.scala:298]: Ok, fair enough. We can always revisit.

**Review by armanbilge — APPROVED** [review]

**Comment by armanbilge** [comment]
I was wrong, better errors is definitely user-facing. Sorry! 😁

**Comment by danicheg** [comment]
It'd be nice if @ChristopherDavenport could glance at this too 🙏🏻

**Comment by armanbilge** [comment]
This seems non-controversial to me, I say we move ahead.

**Pull request merged**